### PR TITLE
add metrics callback

### LINKED ISSUE #7: metric callback
her metric icin bir callback

BaseMetricCallback:
defaulttlar tanimli, labelwise icin extend+overwrite

**on_train/val_batch_end:**
'train'/'val' prefix
metric.add_batch()
(opsiyonel) pl_module.log_any()

**on_train/val_epoch_end:**
'train'/'val' prefix
metric.compute()
pl_module.log_any()

evaluate custom metric destegi: https://huggingface.co/docs/evaluate/main/en/creating_and_sharing

### PR DESCRIPTION
**Pull request by fcakyon**
closes https://github.com/devrimcavusoglu/torchrush-dev/issues/7

metric callback features:
- supports all metrics from hf/evaluate
- supports metric logging on batch end or epoch end
- supports labelwise f1/precision/recall logging
- full compatibility with rushloggers

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,8 +1,9 @@
...
 numpy
-pytorch-lightning==1.8.2
+pytorch-lightning=>1.8.2,<1.9.0
 torchvision==0.13.1
 torch==1.12.1
 tqdm
-evaluate==0.3.0
\ No newline at end of file
+evaluate>=0.3.0,<0.4.0
+scikit-learn>=1.1.3,<1.2.0
```
Comment (devrimcavusoglu) [inline requirements.txt]: `sklearn` eğer sadece metric hesapları içinse (acc, pred, f1) bunu şart koşmamalıyız bence, isteyen kendi projesinde yüklemeli sklearnu kullandığı metrice göre.

File: torchrush/model/base.py [hunk 2]
```diff
@@ -100,20 +101,14 @@ def setup_components(
...
-        Should return dict with key `loss`. Can optionally return keys
-        `predictions` and `references` for metric calculation.
-        """
         pass
 
     @abstractmethod
     def _forward(self, x):
         pass
 
     def forward(self, x):
```
Comment (devrimcavusoglu) [inline torchrush/model/base.py:107]: şu an hala development tarafında olduğumuz için patch gibi görebiliriz bu PRı 😄 bence `_forward()`ı da silip, `forward()`ı abstract yapalım bu durumda çünkü `_forward`a gerek kalmıyor.

File: torchrush/model/lenet5.py [hunk 3]
```diff
@@ -48,10 +49,20 @@ def _forward(self, x):
     def compute_loss(self, y_pred, y_true):
         pass
 
-    def preprocess(self, x):
+    def shared_step(self, batch, batch_idx, mode="train"):
```
Comment (devrimcavusoglu) [inline torchrush/model/lenet5.py:54]: bunu komple silip L53-54'ü forward'a taşıyalım, burada yapmaya gerek yok.

File: torchrush/loggers/__init__.py [hunk 4]
```diff
@@ -1,43 +1,43 @@
-from typing import Any, Dict
+from typing import Any, Dict, Optional
 
 from pytorch_lightning.loggers import CSVLogger, MLFlowLogger, NeptuneLogger, TensorBoardLogger, WandbLogger
```
Comment (devrimcavusoglu) [inline torchrush/loggers/__init__.py]: safe kalmak için (python versiyonları arasında) bu importları şu şekilde değiştirebiliriz.

```suggestion
from pytorch_lightning.loggers import (
  CSVLogger as pl_CSVLogger,
  MLFlowLogger as pl_MLFlowLogger,
  NeptuneLogger as pl_NeptuneLogger,
  TensorBoardLoggeras pl_TensorBoardLogger,
  WandbLogger as pl_WandbLogger
)
```

**Review by devrimcavusoglu — COMMENTED** [review]
Eline sağlık hocam, güzel PR olmuş 👍

File: torchrush/model/lenet5.py (hunk 3 above)
Comment (fcakyon) [reply torchrush/model/lenet5.py:54]: bence forwarddan ayri kalsin bu kisim. 

forward: sadece modelden output almaya yarasin, trainingden bagimsiz sekilde kullanilabilsin, batch'in islenmis halini girdi olarak alsin, cikti olarak sadece raw model ciktisini versin, 

bu shared_step deney sirasinda kullanilan bir wrapper olsun. forwardi cagirsin kendi icinde.

File: requirements.txt (hunk 1 above)
Comment (fcakyon) [reply requirements.txt]: testler icin gerekiyor su anda, test dependencyleri arasina mi ekleyeyim?

File: tests/torchrush/test_metrics.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
+import shutil
```
Comment (fcakyon) [inline tests/torchrush/test_metrics.py:1]: %91 test coverage 😎

File: torchrush/model/lenet5.py [hunk 6]
```diff
@@ -48,10 +49,20 @@ def _forward(self, x):
...
+        # preprocess
+        x, y = batch
         if len(self.input_size) == 1:
```
Comment (devrimcavusoglu) [inline torchrush/model/lenet5.py]: bu ıkı satırı forwarda alabiliriz

**Review by devrimcavusoglu — APPROVED** [review]

**Comment by fcakyon** [comment]
@devrimcavusoglu testleri dedigin sekilde fixture kullanarak guncelledim, wdyt?

8 bagimsiz metric testi oldu boyle, daha iyi oldu 👍

**Review by devrimcavusoglu — APPROVED** [review]

**Comment by devrimcavusoglu** [comment]
Eline sağlık hocam böyle de bence daha güzel olmuş sadece pytest stdouttan bile direkt nerede hata olduğunu görebiliriz daha fine grained şekilde. 👍
